### PR TITLE
UID2-2320 AWS vsock proxy refactoring and cleanup

### PR DESCRIPTION
**Pull request by atarassov-ttd**
- use RAII for memory and other resources management
- simplify channel node management
- use more explicit interfaces and abstractions
- handle less obvious edge cases and error conditions
- change stdout log format to make it easier to follow

This change fixes a few issues along the way:
- initialise variables used to control when unused channels get cleaned up (potential memory leak)
- correctly destroy channel object when returning channel node back to the pool (likely memory leak)
- ensure channel node pool is accessed/updated from a single thread to avoid race conditions (and potential memory leak)
- ensure channel node is returned to the pool if there are errors while setting up channel (potential memory leak)

Tested using:
- local setup with TCP to TCP proxy to nginx and k6 for perf testing
- locally with https://github.com/vi/tcptunnelchecker/ for edge cases around closing connections

### CODE REVIEW

File: vsock-bridge/src/socket.cpp [hunk 1]
```diff
@@ -188,18 +194,14 @@ namespace vsockio
...
-	{
-		if (_inputClosed)
-		{
-			Logger::instance->Log(Logger::DEBUG, "shutdown 6");
-			shutdown();
-		}
+		// There may be a termination buffer queued by the peer. Poller may not be able to detect that and mark
+		// the socket as ready for write in a timely fashion. Force process the queue now.
+		_outputReady = true;
+		writeToOutput();
```
Comment (scong-ttd) [inline vsock-bridge/src/socket.cpp]: I don't get why there is concern if a termination buffer is queued by peer here.

File: vsock-bridge/src/vsock-bridge.cpp [hunk 2]
```diff
@@ -56,20 +56,20 @@ void start_services(std::vector<service_description>& services, int numIOThreads
 
     for (int i = 0; i < numWorkers; i++)
     {
-        WorkerThread* t = new WorkerThread(
+        auto t = std::make_unique<WorkerThread>(
             /*init:*/ []() { 
                 BufferManager::arena->init(512, 2000);
```
Comment (scong-ttd) [inline vsock-bridge/src/vsock-bridge.cpp:61]: Probably should've made it configurable for identity/map scenarios. Out of the scope though.

File: vsock-bridge/src/socket.cpp [hunk 3]
```diff
@@ -208,27 +210,49 @@ namespace vsockio
...
+	{
+		if (!closed())
+		{
+			Logger::instance->Log(Logger::WARNING, "[socket] closing on destruction (fd=", _fd, ")");
+			close();
+		}
+
+		if (_peer != nullptr)
+		{
+			_peer->setPeer(nullptr);
```
Comment (scong-ttd) [inline vsock-bridge/src/socket.cpp:240]: Shall we prevent `Peer::onInputReady` and `Peer::onOutputReady` from panicking from assertion failure if we use `nullptr` to indicate a closed peer?

File: vsock-bridge/src/socket.cpp (hunk 1 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp]: You are right. Force processing the queue from `onPeerClosed()` is sufficient.

File: vsock-bridge/src/socket.cpp (hunk 3 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp:240]: Both sockets should be destroyed together when the parent channel is destroyed. Those callbacks should not be invoked while this is taking place.

File: vsock-bridge/include/buffer.h [hunk 4]
```diff
@@ -1,57 +1,53 @@
...
 namespace vsockio
 {
 	struct MemoryBlock
 	{
-		MemoryBlock(uint8_t* startPtr, class MemoryArena* region)
-			: _startPtr(startPtr), _region(region) {}
-
-		MemoryBlock(MemoryBlock&& other) : _startPtr(other._startPtr), _region(other._region) {}
+		MemoryBlock(int size, class MemoryArena* region)
+			: _startPtr(new uint8_t[size]), _region(region) {}
```
Comment (jon8787) [inline vsock-bridge/include/buffer.h]: we could use make_unique if C++14

File: vsock-bridge/include/channel.h [hunk 5]
```diff
@@ -23,63 +25,61 @@ namespace vsockio
 		int _id;
```
Comment (jon8787) [inline vsock-bridge/include/channel.h:25]: public member variables - intentional?

File: vsock-bridge/src/epoll_poller.cpp [hunk 6]
```diff
@@ -0,0 +1 @@
+#include "epoll_poller.h"
```
Comment (jon8787) [inline vsock-bridge/src/epoll_poller.cpp:1]: curious, why do we need this?

File: vsock-bridge/include/logger.h [hunk 7]
```diff
@@ -24,68 +26,80 @@ struct Logger
 	};
 
 	int _minLevel;
-	static Logger* instance;
+	static Logger *instance;
```
Comment (jon8787) [inline vsock-bridge/include/logger.h]: it was correct before >:(

File: vsock-bridge/include/listener.h [hunk 8]
```diff
@@ -134,56 +145,65 @@ namespace vsockio
...
+                Logger::instance->Log(Logger::ERROR, "creating remote socket failed");
+                return nullptr;
             }
 
-            IOControl::setNonBlocking(fd);
+			auto peer = std::make_unique<Socket>(fd, *SocketImpl::singleton);
+
+            if (!IOControl::setNonBlocking(fd))
+			{
+				const int err = errno;
```
Comment (jon8787) [inline vsock-bridge/include/listener.h]: should we log this err?

File: vsock-bridge/src/socket.cpp [hunk 9]
```diff
@@ -91,30 +103,30 @@ namespace vsockio
 	std::unique_ptr<Buffer> Socket::read()
 	{
 		std::unique_ptr<Buffer> buffer{ BufferManager::getBuffer() };
-		ssize_t bytesRead;
-		ssize_t totalBytes = 0;
+		int bytesRead;
```
Comment (jon8787) [inline vsock-bridge/src/socket.cpp]: we should probably declare `bytesRead` where it's initialized, and totalBytes appears unused now

File: vsock-bridge/src/socket.cpp [hunk 10]
```diff
@@ -91,30 +103,30 @@ namespace vsockio
...
+		int totalBytes = 0;
 
 		while (true)
 		{
-			bytesRead = _impl->read(_fd, buffer->offset(totalBytes), (ssize_t)buffer->capacity() - totalBytes);
+			bytesRead = _impl.read(_fd, buffer->tail(), buffer->remainingCapacity());
 			int err = 0;
```
Comment (jon8787) [inline vsock-bridge/src/socket.cpp:110]: err is unused

File: vsock-bridge/src/socket.cpp [hunk 11]
```diff
@@ -129,56 +141,43 @@ namespace vsockio
...
+		int bytesWritten;
+		while (!buffer.consumed())
 		{
-			ssize_t pageLimit = buffer->pageLimit(totalBytes);
-			ssize_t dataRemaining = buffer->size() - totalBytes;
-			ssize_t lengthToWrite = pageLimit < dataRemaining ? pageLimit : dataRemaining;
-			bytesWritten = _impl->write(_fd, buffer->offset(totalBytes), lengthToWrite);
+			bytesWritten = _impl.write(_fd, buffer.head(), buffer.headLimit());
 
 			int err = 0;
```
Comment (jon8787) [inline vsock-bridge/src/socket.cpp:157]: unused

File: vsock-bridge/src/vsock-bridge.cpp [hunk 12]
```diff
@@ -56,20 +56,20 @@ void start_services(std::vector<service_description>& services, int numIOThreads
...
                 BufferManager::arena->init(512, 2000); 
             }
         );
-        ThreadPool::threads.push_back(t);
+        ThreadPool::threads.push_back(std::move(t));
     }
 
     for (auto& sd : services)
     {
         std::vector<Dispatcher*>* dispatchers = new std::vector<Dispatcher*>();
```
Comment (jon8787) [inline vsock-bridge/src/vsock-bridge.cpp:69]: can we use `vector<unique_ptr<Dispatcher>>` here?

also there's a bug on line 193 and 213 (github won't let me comment there), should be `||`

**Comment by jon8787** [comment]
LGTM but in future we could aim to get rid of more `new`s. Also I wonder if we should be using existing libraries rather than writing our own code for some of these classes.

File: vsock-bridge/include/buffer.h (hunk 4 above)
Comment (atarassov-ttd) [reply vsock-bridge/include/buffer.h]: will fix

File: vsock-bridge/include/channel.h (hunk 5 above)
Comment (atarassov-ttd) [reply vsock-bridge/include/channel.h:25]: Yes for now

File: vsock-bridge/src/epoll_poller.cpp (hunk 6 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/epoll_poller.cpp:1]: To compile the header when building the library. Not strictly needed.

File: vsock-bridge/include/logger.h (hunk 7 above)
Comment (atarassov-ttd) [reply vsock-bridge/include/logger.h]: Fixed

File: vsock-bridge/include/listener.h (hunk 8 above)
Comment (atarassov-ttd) [reply vsock-bridge/include/listener.h]: Removed

File: vsock-bridge/src/socket.cpp (hunk 9 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp]: fixed

File: vsock-bridge/src/socket.cpp (hunk 10 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp:110]: `err` gets assigned/checked in one of the `else if` conditions; and also used in the `else` block. As awkward as it is, it is out of scope for this PR to address.

File: vsock-bridge/src/socket.cpp (hunk 11 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/socket.cpp:157]: Assigned/used in else if and else conditions below.

File: vsock-bridge/src/vsock-bridge.cpp (hunk 12 above)
Comment (atarassov-ttd) [reply vsock-bridge/src/vsock-bridge.cpp:69]: For this PR I focused on repetitive memory allocation/deallocation, not as much on things that get created on startup for the whole lifetime of the app.

Comment (atarassov-ttd) [reply vsock-bridge/src/vsock-bridge.cpp:69]: Fixed the conditionals -- good spotting!

**Review by jon8787 — APPROVED** [review]

**Pull request merged**